### PR TITLE
[BEAM-10027] Support for Kotlin-based Beam Katas

### PR DESCRIPTION
**Pull request by rionmonster**
In this PR, you can find a fork of the latest version of all of the current Java Beam Katas that have been ported to use Kotlin with the associated conversions and dependencies made. 

A summary of the changes can be found below:

- Added appropriate Kotlin-based dependencies for gradle (specific to the Kotlin Katas directory)
- Converted all existing `*.java` tasks files over to Kotlin (e.g. `*.kt`) and updated all of the placeholders where applicable
- Verified that all of the unit tests backing the tasks are passing as expected using the existing input.
- Removed all of the original course metadata (e.g. `*-remote-info.yaml files) assuming that the creation of a new course will overwrite these.

Once the pull-request has been merged (pending any changes), any feedback or recommendations regarding publishing the course to Stepik would be helpful to get the content out to the larger, public audience.

CC: @henryken 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by rionmonster** [comment]
@henryken 

I was able to successfully export the content of this course over to Stepik and generate all of the expected new metadata associated with it (e.g. `*-remote-info.yaml`) files. I have a draft of the course on the site now that as far as I can tell seems to parallel the existing Java one as expected.

I wasn't sure regarding the publication of the actual course itself and what that process entailed (and if it warranted another push to this branch to actually commit the course metadata to support future updates or not). Anyways - I just thought I'd ask. 

Thanks again!

**Comment by henryken** [comment]
Thanks Rion for this additional course in Kotlin!
Would you be able to add me as another author/instructor for this course on Stepik?
You can update it from this [link](https://stepik.org/edit-course-info/72488/).

**Comment by rionmonster** [comment]
Hi Henry!

I’ve added you as both an instructor and author. Thanks so much for your help and for creating the original Beam Katas, as they are a fantastic resource.

File: learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/src/org/apache/beam/learning/katas/windowing/addingtimestamp/withtimestamps/Task.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        val output = applyTransform(events)
+
+        output.apply(Log.ofElements())
+
+        pipeline.run()
+    }
+
+    @JvmStatic
+    fun applyTransform(events: PCollection<Event>): PCollection<Event> {
+        return events.apply(WithTimestamps.of { event: Event -> event.date.toInstant() })
```
Comment (henryken) [inline learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/src/org/apache/beam/learning/katas/windowing/addingtimestamp/withtimestamps/Task.kt:53]: The answer placeholder on Stepik doesn't seem right.
And I also found similar occurrence in few other places.

We can update the course again after the revisions to the PR have been done.

image

**Comment by henryken** [comment]
> Hi @henryken,
> 
> I’ve added you as both an instructor and author. Thanks so much for your help and for creating the original Beam Katas, as they are a fantastic resource.

Thanks @rionmonster! I'm asking JetBrains to help verify the course and make it as featured course.

**Comment by rionmonster** [comment]
Thanks for all the feedback @henryken and the extensive review. I’ll try and get these changes made today and let you know when they are done!

**Comment by henryken** [comment]
Thanks @rionmonster. Take your time. There's no need to rush :)
Enjoy your holiday.

**Comment by rionmonster** [comment]
Hey @henryken, quick question.

What's your opinion on scaffolding out the necessary overrides to solve a particular Kata? Should the placeholder occur within the override itself as in this example:

```
internal class SumIntegerFn : SerializableFunction<Iterable<Int>, Int> {
    override fun apply(input: Iterable<Int>): Int {
        return TODO()
    }
}
```

Or should the user be responsible for writing the override itself as seen below?

```
internal class SumIntegerFn : SerializableFunction<Iterable<Int>, Int> {
    TODO()
}
```

The only reason that I mention this is that for someone new to either Beam or Kotlin, they may not be familiar with the appropriate overrides to apply and the automatic scaffolding from the IDE tends to change the types into their nullable variants, which can be a bit of a hassle to change (e.g. `Int?` as opposed to `Int`).

I just figured I'd ask while I was refactoring through these. 

Thanks,

Rion

File: learning/katas/kotlin/Core Transforms/CoGroupByKey/CoGroupByKey/src/org/apache/beam/learning/katas/coretransforms/cogroupbykey/WordsAlphabet.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.apache.beam.learning.katas.coretransforms.cogroupbykey
+
+class WordsAlphabet(private val alphabet: String?, private val fruit: String, private val country: String) {
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Core Transforms/CoGroupByKey/CoGroupByKey/src/org/apache/beam/learning/katas/coretransforms/cogroupbykey/WordsAlphabet.kt:20]: This change should be made in the latest commit.

**Comment by henryken** [comment]
I think it could be part of the learning to know how to override, if you want to fully understand how to write Beam in Kotlin.
Although I do understand that it could be frustrating if you don't understand Kotlin enough.

I'll leave it up to you to choose which one is better.

**Comment by rionmonster** [comment]
I think I'll keep it in lock step with yours for the time being (e.g. not scaffolding out the overrides), however if we get any flack / feedback on that front, we can adjust it as needed. 

I think outside of that - all of the unused imports have been removed, classes have been converted to data classes where applicable, and the Kotlin Style Guide has been applied to all of the existing Kotlin files within the course (and I went and updated the course after all of these changes).

So I _believe_ that we should be good at this point :)

File: learning/katas/kotlin/Windowing/Adding Timestamp/ParDo/src/org/apache/beam/learning/katas/windowing/addingtimestamp/pardo/Event.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.learning.katas.windowing.addingtimestamp.pardo
+
+import org.joda.time.DateTime
+import java.io.Serializable
+import java.util.*
+
+class Event(private var id: String, private var event: String, var date: DateTime) : Serializable {
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Windowing/Adding Timestamp/ParDo/src/org/apache/beam/learning/katas/windowing/addingtimestamp/pardo/Event.kt]: This change should be made in the latest commit.

File: learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/src/org/apache/beam/learning/katas/windowing/addingtimestamp/withtimestamps/Event.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.learning.katas.windowing.addingtimestamp.withtimestamps
+
+import org.joda.time.DateTime
+import java.io.Serializable
+import java.util.*
+
+class Event(private var id: String, private var event: String, var date: DateTime) : Serializable {
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/src/org/apache/beam/learning/katas/windowing/addingtimestamp/withtimestamps/Event.kt]: This change should be made in the latest commit.

File: learning/katas/kotlin/Windowing/Adding Timestamp/ParDo/task.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+Adding Timestamp - ParDo
+------------------------
+
+Bounded sources (such as a file from TextIO) do not provide timestamps for elements. If you need 
+timestamps, you must add them to your PCollection’s elements.
+
+You can assign new timestamps to the elements of a PCollection by applying a ParDo transform that 
+outputs new elements with timestamps that you set.
+
+**Kata:** Please assign each element a timestamp based on the the `Event.getDate()`.
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Windowing/Adding Timestamp/ParDo/task.md]: Done!

File: learning/katas/kotlin/Core Transforms/Partition/Partition/src/org/apache/beam/learning/katas/coretransforms/partition/Task.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ *  limitations under the License.
+ */
+package org.apache.beam.learning.katas.coretransforms.partition
+
+import org.apache.beam.learning.katas.util.Log
+import org.apache.beam.sdk.Pipeline
+import org.apache.beam.sdk.options.PipelineOptionsFactory
+import org.apache.beam.sdk.transforms.Create
+import org.apache.beam.sdk.transforms.Partition
+import org.apache.beam.sdk.transforms.Partition.PartitionFn
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Core Transforms/Partition/Partition/src/org/apache/beam/learning/katas/coretransforms/partition/Task.kt]: Done!

File: learning/katas/kotlin/Common Transforms/Filter/ParDo/test/org/apache/beam/learning/katas/commontransforms/filter/pardo/TaskTest.kt [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.apache.beam.learning.katas.commontransforms.filter.pardo
+
+import org.apache.beam.learning.katas.commontransforms.filter.pardo.Task.applyTransform
+import org.apache.beam.sdk.testing.PAssert
+import org.apache.beam.sdk.testing.TestPipeline
+import org.apache.beam.sdk.transforms.Create
+import org.apache.beam.sdk.values.PCollection
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Common Transforms/Filter/ParDo/test/org/apache/beam/learning/katas/commontransforms/filter/pardo/TaskTest.kt]: Done!

File: learning/katas/kotlin/Windowing/Fixed Time Window/Fixed Time Window/src/org/apache/beam/learning/katas/windowing/fixedwindow/Task.kt [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.beam.sdk.Pipeline
+import org.apache.beam.sdk.options.PipelineOptionsFactory
+import org.apache.beam.sdk.transforms.Count
+import org.apache.beam.sdk.transforms.Create
+import org.apache.beam.sdk.transforms.windowing.FixedWindows
+import org.apache.beam.sdk.transforms.windowing.Window
+import org.apache.beam.sdk.values.KV
+import org.apache.beam.sdk.values.PCollection
+import org.apache.beam.sdk.values.TimestampedValue
+import org.apache.beam.sdk.values.TypeDescriptors
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Windowing/Fixed Time Window/Fixed Time Window/src/org/apache/beam/learning/katas/windowing/fixedwindow/Task.kt]: Done!

File: learning/katas/kotlin/Examples/Word Count/Word Count/src/org/apache/beam/learning/katas/examples/wordcount/Task.kt [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+package org.apache.beam.learning.katas.examples.wordcount
+
+import org.apache.beam.learning.katas.util.Log
+import org.apache.beam.sdk.Pipeline
+import org.apache.beam.sdk.options.PipelineOptionsFactory
+import org.apache.beam.sdk.transforms.*
+import org.apache.beam.sdk.values.KV
+import org.apache.beam.sdk.values.PCollection
+import org.apache.beam.sdk.values.TypeDescriptors
+import java.util.*
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Examples/Word Count/Word Count/src/org/apache/beam/learning/katas/examples/wordcount/Task.kt]: Done!

File: learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/src/org/apache/beam/learning/katas/windowing/addingtimestamp/withtimestamps/Task.kt (hunk 1 above)
Comment (rionmonster) [reply learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/src/org/apache/beam/learning/katas/windowing/addingtimestamp/withtimestamps/Task.kt:53]: This should be adjusted; I tested it through and it worked locally.

File: learning/katas/kotlin/Common Transforms/Filter/ParDo/src/org/apache/beam/learning/katas/commontransforms/filter/pardo/Task.kt [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        output.apply(Log.ofElements())
+
+        pipeline.run()
+    }
+
+    @JvmStatic
+    fun applyTransform(input: PCollection<Int>): PCollection<Int> {
+        return input.apply(ParDo.of(object : DoFn<Int, Int>() {
+                    @ProcessElement
+                    fun processElement(@Element element: Int, context: ProcessContext, out: OutputReceiver<Int>) {
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Common Transforms/Filter/ParDo/src/org/apache/beam/learning/katas/commontransforms/filter/pardo/Task.kt]: Done!

File: learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/test/org/apache/beam/learning/katas/windowing/addingtimestamp/withtimestamps/TaskTest.kt [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.beam.sdk.transforms.Create
+import org.apache.beam.sdk.transforms.DoFn
+import org.apache.beam.sdk.transforms.ParDo
+import org.apache.beam.sdk.values.KV
+import org.joda.time.DateTime
+import org.joda.time.Instant
+import org.junit.Rule
+import org.junit.Test
+import java.io.Serializable
+import java.util.*
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/test/org/apache/beam/learning/katas/windowing/addingtimestamp/withtimestamps/TaskTest.kt]: Done!

File: learning/katas/kotlin/Core Transforms/Map/FlatMapElements/src/org/apache/beam/learning/katas/coretransforms/map/flatmapelements/Task.kt [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+import org.apache.beam.learning.katas.util.Log
+import org.apache.beam.sdk.Pipeline
+import org.apache.beam.sdk.options.PipelineOptionsFactory
+import org.apache.beam.sdk.transforms.Create
+import org.apache.beam.sdk.transforms.FlatMapElements
+import org.apache.beam.sdk.transforms.SerializableFunction
+import org.apache.beam.sdk.values.PCollection
+import org.apache.beam.sdk.values.TypeDescriptors
+import java.util.*
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Core Transforms/Map/FlatMapElements/src/org/apache/beam/learning/katas/coretransforms/map/flatmapelements/Task.kt]: Done!

File: learning/katas/kotlin/Core Transforms/Combine/CombineFn/src/org/apache/beam/learning/katas/coretransforms/combine/combinefn/Task.kt [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.beam.learning.katas.coretransforms.combine.combinefn.Task.AverageFn.Accum
+import org.apache.beam.learning.katas.util.Log
+import org.apache.beam.sdk.Pipeline
+import org.apache.beam.sdk.options.PipelineOptionsFactory
+import org.apache.beam.sdk.transforms.Combine
+import org.apache.beam.sdk.transforms.Combine.CombineFn
+import org.apache.beam.sdk.transforms.Create
+import org.apache.beam.sdk.values.PCollection
+import java.io.Serializable
+import java.util.*
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Core Transforms/Combine/CombineFn/src/org/apache/beam/learning/katas/coretransforms/combine/combinefn/Task.kt:29]: Done!

**Comment by rionmonster** [comment]
@henryken 

I noticed that I had missed a few of the issues that you had mentioned earlier in the chain for this issue. I believe that everything should up addressed, so just let me know if there's anything else you think we'd need to do.

Thanks!

Rion

File: learning/katas/kotlin/Core Transforms/Combine/CombineFn/src/org/apache/beam/learning/katas/coretransforms/combine/combinefn/Task.kt [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @JvmStatic
+    fun applyTransform(input: PCollection<Int>): PCollection<Double> {
+        return input.apply(Combine.globally(AverageFn()))
+    }
+
+    internal class AverageFn : CombineFn<Int, Accum, Double>() {
+
+        internal inner class Accum : Serializable {
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Core Transforms/Combine/CombineFn/src/org/apache/beam/learning/katas/coretransforms/combine/combinefn/Task.kt]: Done

File: learning/katas/kotlin/util/test/org/apache/beam/learning/katas/util/ContainsKvs.kt [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+            return ContainsKvs(ImmutableList.copyOf(kvs))
+        }
+    }
+
+    override fun apply(input: Iterable<KV<String, Iterable<String>>>): Void? {
+        val matchers: MutableList<Matcher<in KV<String, Iterable<String>>>> = ArrayList()
+
+        for (expected in expectedKvs) {
+            val values = Iterables.toArray(expected.value, String::class.java)
+            matchers.add(KvMatcher.Companion.isKv(CoreMatchers.equalTo(expected.key), IsIterableContainingInAnyOrder.containsInAnyOrder(*values)))
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/util/test/org/apache/beam/learning/katas/util/ContainsKvs.kt]: Done!

File: learning/katas/kotlin/Windowing/Fixed Time Window/Fixed Time Window/test/org/apache/beam/learning/katas/windowing/fixedwindow/WindowedEvent.kt [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.learning.katas.windowing.fixedwindow
+
+import java.io.Serializable
+import java.util.*
+
+class WindowedEvent(private val event: String?, private val count: Long, private val window: String) : Serializable {
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Windowing/Fixed Time Window/Fixed Time Window/test/org/apache/beam/learning/katas/windowing/fixedwindow/WindowedEvent.kt]: Done!

**Comment by rionmonster** [comment]
@henryken

Reading over the initial PR request (and given the size of this thing), do you think it warrants filing an ICLA?

**Comment by henryken** [comment]
For the ICLA, I think it is not required. I didn't sign it previously when submitting the Beam Katas work.

**Comment by rionmonster** [comment]
@henryken

That’s interesting, I had applied the appropriate style guide, which also handled the removal of unused imports as well as updated all of the placeholders.

I’ll run back through them and make all of the appropriate changes today to get this taken care of.

File: learning/katas/kotlin/Windowing/Fixed Time Window/Fixed Time Window/test/org/apache/beam/learning/katas/windowing/fixedwindow/WindowedEvent.kt [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.learning.katas.windowing.fixedwindow
+
+import java.io.Serializable
+import java.util.*
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Windowing/Fixed Time Window/Fixed Time Window/test/org/apache/beam/learning/katas/windowing/fixedwindow/WindowedEvent.kt]: Done!

File: learning/katas/kotlin/Core Transforms/Combine/CombineFn/src/org/apache/beam/learning/katas/coretransforms/combine/combinefn/Task.kt [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.beam.learning.katas.coretransforms.combine.combinefn.Task.AverageFn.Accum
+import org.apache.beam.learning.katas.util.Log
+import org.apache.beam.sdk.Pipeline
+import org.apache.beam.sdk.options.PipelineOptionsFactory
+import org.apache.beam.sdk.transforms.Combine
+import org.apache.beam.sdk.transforms.Combine.CombineFn
+import org.apache.beam.sdk.transforms.Create
+import org.apache.beam.sdk.values.PCollection
+import java.io.Serializable
+import java.util.*
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Core Transforms/Combine/CombineFn/src/org/apache/beam/learning/katas/coretransforms/combine/combinefn/Task.kt]: Done!

File: learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/task.md [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+Adding Timestamp - WithTimestamps
+---------------------------------
+
+Bounded sources (such as a file from TextIO) do not provide timestamps for elements. If you need 
+timestamps, you must add them to your PCollection’s elements.
+
+You can assign new timestamps to the elements of a PCollection by applying a ParDo transform that 
+outputs new elements with timestamps that you set.
+
+**Kata:** Please assign each element a timestamp based on the the `Event.getDate()`.
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Windowing/Adding Timestamp/WithTimestamps/task.md]: Done!

File: learning/katas/kotlin/Core Transforms/Map/MapElements/task-info.yaml [hunk 20]
```diff
@@ -0,0 +1,13 @@
+type: edu
+files:
+- name: src/org/apache/beam/learning/katas/coretransforms/map/mapelements/Task.kt
+  visible: true
+  placeholders:
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Core Transforms/Map/MapElements/task-info.yaml:5]: Done!

File: learning/katas/kotlin/Core Transforms/Combine/CombineFn/task-info.yaml [hunk 21]
```diff
@@ -0,0 +1,10 @@
+type: edu
+files:
+- name: src/org/apache/beam/learning/katas/coretransforms/combine/combinefn/Task.kt
+  visible: true
+  placeholders:
+  - offset: 1951
```
[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Core Transforms/Combine/CombineFn/task-info.yaml]: Done!

**Comment by rionmonster** [comment]
@henryken 

I just pushed up another commit that should apply all of the styling-related changes that you requested (specifically 2-space indentions), however I wasn't able to get the CheckStyle approach to work (I simply used the preferred Kotlin Coding Style from Kotlin and adjusted it to use 2-space indention as opposed to 4).

Additionally, I double-checked the placeholders that you had mentioned and addressed any lingering, unused imports as well.

Thanks,

Rion

File: learning/katas/kotlin/Core Transforms/Map/MapElements/task-info.yaml (hunk 20 above)
[the remark this answers is not in the document]
Comment (henryken) [reply learning/katas/kotlin/Core Transforms/Map/MapElements/task-info.yaml:5]: Can you exclude the "return" from the answer placeholder, similar to the other tasks in the Map lesson?
image

[the remark this answers is not in the document]
Comment (rionmonster) [reply learning/katas/kotlin/Core Transforms/Map/MapElements/task-info.yaml:5]: Done (and checked through the remaining Map tasks to ensure they aligned).

**Comment by rionmonster** [comment]
@henryken / @pabloem 

I just pushed up the changes you mentioned regarding the Map placeholders, so they should be good to go now. I've also updated the course on Stepik (and included those updates within the commit as well).

I think we are good to go now. Thanks for all of your help Henry!

Rion

**Comment by henryken** [comment]
Thanks @rionmonster!

@pabloem, please help to merge.

**Comment by pabloem** [comment]
woah very exciting!

**Comment by pabloem** [comment]
Thanks @rionmonster ! - just need to get RAT validations passing.

look at files marked with missing apache license headers: https://builds.apache.org/job/beam_PreCommit_RAT_Commit/17932/console

They should be skipped from the validation, or have a header added. @henryken what have you done for others?

**Comment by rionmonster** [comment]
@pabloem / @henryken 

I just sent up another commit that added the appropriate licensing headers to all of the `task-info.yaml` files, which should align these Kotlin ones with those found within Henry's Katas.

Hopefully that'll help!

Rion

**Comment by pabloem** [comment]
okay let me trigger the tests... (by asking Jenkins...)

**Comment by pabloem** [comment]
retest this please

**Comment by rionmonster** [comment]
@pabloem

Doh! It looks like one of the headers was missed. Just pushed up another commit to add that one.

**Comment by rionmonster** [comment]
retest this please

**Comment by pabloem** [comment]
retest this please

**Comment by pabloem** [comment]
unfortunately the ASF now only allows committers to trigger tests. Triggering...

**Comment by pabloem** [comment]
Thank you!

**Comment by pabloem** [comment]
thanks a lot @rionmonster @henryken - I'd encourage you to announce on user@ or even write a blog post if you feel up to it : )

**Comment by rionmonster** [comment]
@henryken / @pabloem 

Thanks to both of you guys for helping me through this process, both in review and getting it merged in. I shot out an e-mail on the user mailing list, and will likely write up a short blog post on my personal blog when I get a chance. If there's a more appropriate place to add links to the course, or another PR I could push up to the Beam site, please let me know and I can put together something for that as well.

I know I've personally run into quite a few "gotchas" related to iterations between Beam and Kotlin, so it wouldn't hurt to document those somewhere (e.g. a blog post) as well.

Thanks again guys!

**Comment by henryken** [comment]
We've made it! 🙌

**Pull request merged**